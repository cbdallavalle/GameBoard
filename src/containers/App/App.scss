.App {
  text-align: center;
  background-image: url('../../assets/risk.jpg');
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  min-height: 100vh;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

#app-error {
  background-color: #ff38389c;
  width: 400px;
  color: #5f0404;
  margin: auto;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}