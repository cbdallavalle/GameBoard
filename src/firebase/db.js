import { db } from './firebase';

export const doCreateUser = (id, firstName, lastName, email) =>
  db.ref(`users/${id}`).set({
    user: {
      id,
      firstName,
      lastName,
      email,
    }
  });

export const onceGetUsers = async () => {
  try {
    const snapshot = await db.ref('users').once('value');
    if (snapshot.status < 300) {
      const value = await snapshot.val();
      return value
    } else {
      throw new Error('unable to fetch users')
    }
  } catch (error) {
    throw new Error('unable to fetch users')
  }
}

export const doWriteFavoriteData = async (userId, newFavorites) => {
  db.ref('users/' + userId + '/favorites').set(newFavorites);
}

export const doAddFavoriteData = async (userId, favorite) => {
  const existingFavorites = await getFavorites(userId);
  const favoriteToAdd = checkFavoriteDuplication(existingFavorites, favorite);
  doWriteLastFavoritedData(userId, favorite)
  const newFavorites = {...existingFavorites, ...favoriteToAdd }
  this.doWriteFavoriteData(userId, newFavorites)
}

export const doDeleteFavoriteData = async (userId, favorite) => {
  const existingFavorites = await getFavorites(userId);
  delete existingFavorites[favorite.name]
  this.doWriteFavoriteData(userId, existingFavorites)
}

export const doWriteLastFavoritedData = async (userId, favorite) => {
  db.ref('users/' + userId + '/lastFavorited').set(favorite);
}

export const checkFavoriteDuplication = (existingFavorites, favorite) => {
  const { name, id, description, image, thumbnail } = favorite;
  const review = "No review - click to add one";
  const favoriteToAdd = { [name]: { id, description, image, thumbnail, name, review } }
  return existingFavorites && existingFavorites[favorite.name] ? {} : favoriteToAdd
}

export const doWriteFriendsData = async (userId, friendId) => {
  try {
    const friends = await getFriends(userId);
    if (friends && !friends.includes(friendId)) {
      friends.push(friendId);
      db.ref('users/' + userId + '/friends').set(friends); 
    } else if (!friends) {
      db.ref('users/' + userId + '/friends').set([friendId]); 
    }
  } catch (error) {
    throw new Error ('unable to add friend')
  }
}

export const getFriends = async (userId) => {
  const snapshot = await db.ref('users').once('value');
  const value = await snapshot.val();
  return value[userId].friends
}

export const getFavorites = async (userId) => {
  const value = await onceGetUsers();
  return value[userId].favorites;
}

export const getFriendsFavorites = async (userId) => {
  const friends = await getFriends(userId);
  const value = await onceGetUsers();
  if(friends) {
    return friends.reduce( (accu, friendId) => {
      if(value[friendId].lastFavorited) {
        accu[value[friendId].user.firstName] = value[friendId].lastFavorited
      }
      return accu
    }, {} )  
  } else {
    return {}
  }
}

export const doWriteReviewData = async (userId, favorite, review) => {
  const { name } = favorite;
  const reviewedGame = {...favorite, review}
  db.ref('users/' + userId + '/favorites/' + name).set(reviewedGame);
  await doWriteLastFavoritedData(userId, reviewedGame)
}