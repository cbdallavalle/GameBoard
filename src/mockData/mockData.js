export const mockData = {
  mockUsers: {
    YRrtMzWo0jNquYjLCQGT8ftqz703:{
      favorites: {}, 
      friends: [], 
      lastFavorited: {}, 
      user: {
        email: "kubla@email.com",
        firstName: "kubla",
        id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
        lastName: "khan"
      }
    },
    GZZGrEW9aYXzbxrdphXuq04ng6E3:{
      favorites: {}, 
      friends: [], 
      lastFavorited: {}, 
      user: {
        email: "wren@wren.com",
        firstName: "wren",
        id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
        lastName: "little" 
      }
    }
  },
  mockAllUsersWithoutSearcher: [
    {
      email: "kubla@email.com", 
      firstName: "kubla", 
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3", 
      lastName: "khan", 
      key: "YRrtMzWo0jNquYjLCQGT8ftqz703"
    }
  ],
  mockAllUsers: [
    {
      email: "kubla@email.com", 
      firstName: "kubla", 
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3", 
      lastName: "khan", 
      key: "YRrtMzWo0jNquYjLCQGT8ftqz703"
    },
    {
      email: "wren@wren.com",
      firstName: "wren",
      id: "GZZGrEW9aYXzbxrdphXuq04ng6E3",
      lastName: "little" 
    }
  ],
  mockFriends: ['123', '456', '778'],
  mockFavorites: { 'Race for the Galaxy': {game: 'info'} },
  mockFriendsFavorites: {
    kubla: { 'Race for the Galaxy': {game: 'info'} },
    wren: { 'Captain Sonar': {game: 'info'} }
  },
  initialState: {
    firstName: '',
    lastName: '',
    email: '',
    passwordOne: '',
    passwordTwo: '',
    error: null,
    displayCreate: 'login'
  },
  mockLogInEvent: {
    target: {
      name: 'email', 
      value: 'coda@gmail.com'
    }, 
    preventDefault: () => {}
  },
  mockDefaultSearchState: {
    games: [],
    search: '',
    game: {},
    error: '',
    loading: false,
    gameAdded: 'not-added'
  },
  mockSearchEvent: {
    target: {
      name: 'search',
      value: 'Meeple Circus'
    },
    preventDefault: () => {}
  }, 
  mockSearch: [{ 
    id: '123', 
    name: 'Mysterium' 
  }],
  mockSearchGame: { 
    id: '123', 
    name: 'Mysterium', 
    thumbnail: 'src', 
    image: 'src', 
    description: 'A mystery game with pictures' 
  },
  mockMSTPUserState: {
    user: {
      uid: 'GZZGrEW9aYXzbxrdphXuq04ng6E3'
    }
  },
  mockMSTPUser: {
    user: {
      uid: 'GZZGrEW9aYXzbxrdphXuq04ng6E3'
    }
  }
};