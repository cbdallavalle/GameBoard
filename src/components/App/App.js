import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import Main from '../Main/Main';
import { auth } from '../../firebase';
import { loginUser } from '../../actions';
import './App.css';

export class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authUser: null
    };
  }

  componentDidMount = async () => {
    await auth.onAuthStateChanged(authUser => {
      authUser
        ? this.props.loginUser(authUser)
        : this.props.loginUser(null)
    });
  }
  
  render() {
    return (
      <div className="App">
        <Main />
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loginUser: user => dispatch(loginUser(user))
})

export default withRouter(connect(null, mapDispatchToProps)(App));
