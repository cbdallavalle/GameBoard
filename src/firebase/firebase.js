import * as firebase from 'firebase';

const config = {
  apiKey: "AIzaSyBDU0_Pa8xdAWKmtKKfo3QkbrtR56cl024",
  authDomain: "dallavalle-gameboard.firebaseapp.com",
  databaseURL: "https://dallavalle-gameboard.firebaseio.com",
  projectId: "dallavalle-gameboard",
  storageBucket: "",
  messagingSenderId: "250758219123"
};

if(!firebase.apps.length) {
  firebase.initializeApp(config)
}

const auth = firebase.auth();

export {
  auth,
}