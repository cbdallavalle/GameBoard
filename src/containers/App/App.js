import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import Main from '../Main/Main';
import { auth, db } from '../../firebase';
import { loginUser, updateFavorites, updateFriends } from '../../actions';
import './App.css';

export class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authUser: null,
      error: ''
    };
  }

  componentDidMount = async () => {
    await auth.onAuthStateChanged(authUser => {
      authUser
        ? this.loginUser(authUser)
        : this.props.loginUser(null);
    });
  }

  loginUser = async(authUser) => {
    const user = {uid: authUser.uid};
    this.props.loginUser(user);
    this.updateFriends(user.uid);
    this.updateFavorites(user.uid);
  }

  updateFriends = async(userId) => {
    try {
      const friends = await db.getFriends(userId);
      friends && this.props.updateFriends(friends);
    } catch (error) {
      this.setState({ error: error.message });
    }
  }

  updateFavorites = async(userId) => {
    try {
      const favorites = await db.getFavorites(userId);
      favorites && this.props.updateFavorites(favorites);
    } catch (error) {
      this.setState({ error: error.message });
    }
  }
  
  render() {
    return (
      <div className="App">
        <p id="app-error">{ this.state.error }</p>
        <Main />
      </div>
    );
  }
}

App.propTypes = {
  loginUser: PropTypes.func.isRequired,
  updateFavorites: PropTypes.func.isRequired,
  updateFriends: PropTypes.func.isRequired
};

export const mapDispatchToProps = dispatch => ({
  loginUser: user => dispatch(loginUser(user)),
  updateFavorites: favorites => dispatch(updateFavorites(favorites)),
  updateFriends: friends => dispatch(updateFriends(friends))
});

export default withRouter(connect(null, mapDispatchToProps)(App));