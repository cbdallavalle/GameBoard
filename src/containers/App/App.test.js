import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapDispatchToProps } from './App';
import { db } from '../../firebase';
import { shallow } from 'enzyme';
import { mockData } from '../../mockData/mockData';

describe("App", () => {
  const mockLoginUser = jest.fn();
  const mockUpdateFavorites = jest.fn();
  const mockUpdateFriends = jest.fn();

  it('should match the snapshot', () => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should have a default state of authUser of null', () => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    , { disableLifecycleMethods: true })
    expect(wrapper.state()).toEqual({ authUser: null, error: '' })
  })

  it('should run loginUser on component did mount', () => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    expect(mockLoginUser).toHaveBeenCalled();
  })

  //loginUser
  it('loginUser should call loginUser, updateFriends, and updateFavorites', () => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    wrapper.instance().updateFriends = jest.fn();
    wrapper.instance().updateFavorites = jest.fn();
    wrapper.instance().loginUser({uid: '1234'});
    expect(mockLoginUser).toHaveBeenCalledWith({uid: '1234'});
    expect(wrapper.instance().updateFriends).toHaveBeenCalledWith('1234');
    expect(wrapper.instance().updateFavorites).toHaveBeenCalledWith('1234');
  })

  //updateFriends
  it('should set state with an error if error is caught', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )

    window.fetch = jest.fn().mockImplementation( () => Promise.resolve({
      status: 500,
      error: new Error('unable to load friends')
    }))

    await wrapper.instance().updateFriends();
    expect(wrapper.state().error).toEqual('unable to load friends data')
  })

  it('updateFriends should call mockUpdateFriends if friends exist', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    db.getFriends = () => mockData.mockFriends;
    await wrapper.instance().updateFriends('123');
    expect(mockUpdateFriends).toHaveBeenCalledWith(mockData.mockFriends);
  })

  it('updateFriends should not call mockUpdateFriends if friends does not exist', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    wrapper.instance().props.updateFriends.mockClear()
    db.getFriends = (userId) => undefined;
    await wrapper.instance().updateFriends('123');
    expect(mockUpdateFriends).not.toHaveBeenCalled();
  })

  //updateFavorites
  it('updateFavorites should set state with an error if error is caught', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )

    window.fetch = jest.fn().mockImplementation( () => Promise.resolve({
      status: 500,
      error: new Error('unable to load favorites data')
    }))

    await wrapper.instance().updateFavorites();
    expect(wrapper.state().error).toEqual('unable to load favorites data')
  })

  it('updateFavorites should call mockUpdateFavorites if favorites exist', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    db.getFavorites = () => mockData.mockFavorites;
    await wrapper.instance().updateFavorites('123');
    expect(mockUpdateFavorites).toHaveBeenCalledWith(mockData.mockFavorites);
  })

  it('updateFavorites should not call mockUpdateFavorites if favorites does not exist', async() => {
    const wrapper = shallow(
      <App 
        loginUser={mockLoginUser}
        updateFavorites={mockUpdateFavorites}
        updateFriends={mockUpdateFriends}
      />
    )
    wrapper.instance().props.updateFavorites.mockClear()
    db.getFavorites = (userId) => undefined;
    await wrapper.instance().updateFavorites('123');
    expect(mockUpdateFavorites).not.toHaveBeenCalled();
  })
})

describe("mapDispatchToProps", () => {
  
  it('should call dispatch when loginUser is called', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.loginUser();
    expect(mockDispatch).toHaveBeenCalled();    
  })

  it('should call dispatch when updateFavorites is called', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.updateFavorites();
    expect(mockDispatch).toHaveBeenCalled();    
  })

  it('should call dispatch when updateFriends is called', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.updateFriends();
    expect(mockDispatch).toHaveBeenCalled();    
  })
})
