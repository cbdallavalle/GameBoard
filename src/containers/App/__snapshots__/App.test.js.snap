// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    loginUser={[Function]}
    updateFavorites={[Function]}
    updateFriends={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p />,
        <withRouter(Connect(Component)) />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p />,
          <withRouter(Connect(Component)) />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
